### PR TITLE
uv: export uv_backend_fd nad uv_backend_timeout

### PR DESCRIPTION
**Pull request by indutny**
This could be used with uv_run_once() to poll in one thread and run
event-loop's thread's event callbacks in another. Which is useful for
embedding uv's event-loop in another event-loop.

/cc @bnoordhuis

### CODE REVIEW

File: include/uv.h [hunk 1]
```diff
@@ -261,6 +261,17 @@
 UV_EXTERN void uv_update_time(uv_loop_t*);
 UV_EXTERN int64_t uv_now(uv_loop_t*);
 
+/*
+ * Get backend file descriptor (only kqueue and epoll are supported),
```
Comment (bnoordhuis) [inline include/uv.h]: And event ports?

File: include/uv.h [hunk 2]
```diff
@@ -261,6 +261,17 @@
...
+ * and get polling timeout for it.
+ *
+ * This could be used with uv_run_once() to poll in one thread and run
+ * event-loop's thread's event callbacks in another. Which is useful for
+ * embedding uv's event-loop in another event-loop.
```
Comment (bnoordhuis) [inline include/uv.h]: s/event-loop/event loop/g

File: include/uv.h [hunk 3]
```diff
@@ -261,6 +261,17 @@
...
+ */
+UV_EXTERN int uv_backend_fd(uv_loop_t*);
+UV_EXTERN unsigned int uv_backend_timeout(uv_loop_t*);
```
Comment (bnoordhuis) [inline include/uv.h]: `const uv_loop_t*`?

EDIT: That'll only work for uv_backend_fd().

File: test/test-embed.c [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+static uv_thread_t embed_thread;
+static uv_sem_t embed_sem;
+static uv_timer_t embed_timer;
+static uv_async_t embed_async;
+static volatile int embed_closed;
+
+static int embed_timer_called;
+
+
+static void embed_thread_runner(void *arg) {
```
Comment (bnoordhuis) [inline test/test-embed.c]: Style.

File: test/test-embed.c [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  while (!embed_closed) {
+    int r;
+    int fd;
+    unsigned int timeout;
+    struct timespec ts;
```
Comment (bnoordhuis) [inline test/test-embed.c]: Declarations go at the top.

File: test/test-embed.c [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+#include <stdio.h>
+#include <stdlib.h>
+
+#if defined(HAVE_KQUEUE) || defined(HAVE_EPOLL)
```
Comment (bnoordhuis) [inline test/test-embed.c]: Where are these defined?

File: test/test-embed.c [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    fd = uv_backend_fd(uv_default_loop());
+    timeout = uv_backend_timeout(uv_default_loop());
+
+    do {
+#if defined(HAVE_KQUEUE)
+      ts.tv_sec = timeout / 1000;
+      ts.tv_nsec = (timeout % 1000) * 1000000;
+      r = kevent(fd, NULL, 0, NULL, 0, timeout == -1 ? NULL : &ts);
+#elif defined(HAVE_EPOLL)
+      r = uv__epoll_wait(fd, NULL, 0, timeout);
```
Comment (bnoordhuis) [inline test/test-embed.c]: libuv internal. I'm 99% sure this won't compile with -Werror.

File: test/test-embed.c (hunk 4 above)
Comment (indutny) [inline test/test-embed.c]: huh?

Comment (indutny) [inline test/test-embed.c]: ah, void *

**Comment by saghul** [comment]
I guess this should give UV_ENOSYS on Windows, right?

File: include/uv.h [hunk 8]
```diff
@@ -261,6 +261,17 @@
...
+/*
+ * Get backend file descriptor and get polling timeout for it.
+ * (only kqueue, epoll and event ports are supported).
+ *
+ * This could be used with uv_run_once() to poll in one thread and run
+ * event loop's thread's event callbacks in another. Which is useful for
+ * embedding uv's event loop in another event loop.
+ */
+UV_EXTERN int uv_backend_fd(const uv_loop_t*);
+UV_EXTERN unsigned int uv_backend_timeout(uv_loop_t*);
```
Comment (bnoordhuis) [inline include/uv.h:273]: Needs no-op win32 implementations, otherwise LGTM.

**Comment by indutny** [comment]
@bnoordhuis done

**Comment by bnoordhuis** [comment]
Thanks Fedor, landed in 09a7f85. I updated uv_backend_timeout()'s prototype in 4a69c4b.

**Comment by deanm** [comment]
Just want to comment that runloop integration is something I've tackled a few times for Plask.  I've just recently rewritten it for libuv and taken a new approach (that is kqueue specific).  There are at least some helpful comments that explain the approach and might be helpful:

https://github.com/deanm/plask/commit/3db9c663adb1d24bc07e523614fef8e4167cdbc3

**Pull request closed without merging**